New CLI options, custom tags range and tests.

emmenko opened the pull request:
Closes #17, #19, #25

### Summary
This PR mainly introduces a way of querying commits for a custom tags range (at the moment only commits after the last available tag are taken into consideration) and to group them by tag.

### Description
> There are quite some changes in the diff (`Changelog.js`), most of them are just about formatting and code improvements though. I'll point out in the diff the places to look at, regarding the actual changes.
Those changes also should not break anything, as far as I can tell the old behaviour is still preserved.

1. CLI options
To be able to query the commits for a custom range of tags, there are 2 options that we can use:

```bash
--tag-from <tag> # default is the latest available tag
--tag-to <tag>

# --tag-to should be a tag after --tag-from
```

---
<details>
<summary>Slightly different approach by using "positional arguments"</summary>
It's kind of a workaround because `yargs` supports that only for _commands_ and because if we want them to be `<fromTag>..<toTag>` it's considered as 1 argument (because of the `..`) and we need to do some parsing anyway.

```bash
$ lerna-changelog <fromTag>..<toTag>
```

```diff
diff --git cli.js cli.js
index 841e66a..ae2d5c0 100755
--- cli.js
+++ cli.js
@@ -1,39 +1,57 @@
 #!/usr/bin/env node
-
 var chalk = require("chalk");
 var Changelog = require(".").Changelog;
 var ConfigurationError = require(".").ConfigurationError;

 var argv = require("yargs")
-  .usage("Usage: lerna-changelog [options]")
-  .options({
-    "tag-from": {
-      type: "string",
-      desc: "A git tag that determines the lower bound of the range of commits (defaults to last available)"
-    },
-    "tag-to": {
-      type: "string",
-      desc: "A git tag that determines the upper bound of the range of commits"
-    }
-  })
+  .usage("Usage: lerna-changelog <fromTag>..<toTag>")
   .example(
     "lerna-changelog",
     "create a changelog for the changes after the latest available tag"
   )
   .example(
-    "lerna-changelog --tag-from 0.1.0 --tag-to 0.3.0",
+    "lerna-changelog 0.1.0",
+    "create a changelog for the changes after the given tag"
+  )
+  .example(
+    "lerna-changelog 0.1.0..0.3.0",
     "create a changelog for the changes in all tags within the given range"
   )
   .version()
   .help()
   .argv;

+var fromTag;
+var toTag;
+var args = argv._;
+if (args.length > 1) {
+  console.error(
+    chalk.red("Invalid arguments, must be in format `<fromTag>..<toTag>`")
+  );
+  process.exit(1);
+} else if (args.length === 1) {
+  var indexOfSeparator = args[0].indexOf("..");
+  if (indexOfSeparator === 0) {
+    console.error(
+      chalk.red("The lower bound tag is required `<fromTag>..<toTag>`")
+    );
+    process.exit(1);
+  } else if (indexOfSeparator === -1) {
+    fromTag = args[0];
+  } else {
+    var parts = args[0].split("..");
+    fromTag = parts[0];
+    toTag = parts[1];
+  }
+}
+
+var options = { fromTag: fromTag, toTag: toTag };
 try {
-  console.log((new Changelog(argv)).createMarkdown());
+  console.log(new Changelog(options).createMarkdown());
 } catch (e) {
   if (e instanceof ConfigurationError) {
     console.log(chalk.red(e.message));
   } else {
-    throw (e);
+    throw e;
   }
 }
```
</details>

---

2. Lerna.json options
A new option `ignoreCommitters` can be used in `lerna.json` to ignore matching committers. This is particularly useful for commiters such as `greenkeeper` or any kind of bot accounts.

3. Changelog entries grouped by tag
As briefly discussed in #25, having a list of commits possibly spread across multiple tags should result in multiple changelog entries for each tag.

**Empty tags (with no commits) will be discarded** 

```
## Unreleased (2017-01-15)

#### :rocket: New Feature
* `pkg-1`, `pkg-2`
  * This is the commit title for the issue (#4). ([@test-user](https://github.com/test-user))
  * This is the commit title for the issue (#3). ([@test-user](https://github.com/test-user))

#### Committers: 1
- Test User ([test-user](https://github.com/test-user))


## v0.2.0 (2017-01-01)

#### :rocket: New Feature
* `pkg-1`,
  * This is the commit title for the issue (#2). ([@test-user](https://github.com/test-user))
  * This is the commit title for the issue (#1). ([@test-user](https://github.com/test-user))

#### Committers: 1
- Test User ([test-user](https://github.com/test-user))
```

4. Tests
Some basic tests have been added, including mocked modules.

### TODO
- [ ] setup travis build
- [ ] document how changelog is generated (closes #30)
- [ ] make npm aware of executables (move `cli.js` into `bin` folder)
- [ ] `--summary` option to print information about processed commits (if commits got discarded, etc)

emmenko commented on package.json:
```diff
@@ -7,9 +7,10 @@
   },
   "scripts": {
-    "build": "babel src -d lib",
+    "build": "npm run clean && babel src --out-dir lib --ignore src/__mocks__/GithubAPI.js,src/__mocks__/Changelog.js",
```
I'm not sure if there is a better way of doing this...

hzoo replied:
We can just ignore in npmignore/files? (I normally but all tests in a test folder not like jest so not sure)

emmenko replied:
The problem is I keep those mocked folders is that they will be present in `/lib` as well as `/src` and when `jest` runs it outputs some logs. Unfortunately I couldn't find a way of configuring this with `jest`, if you know of a better solution I'm all ears.

```bash
jest-haste-map: duplicate manual mock found:
  Module name: GithubAPI
  Duplicate Mock path: /Users/emmenko/dev/src/lerna-changelog/src/__mocks__/GithubAPI.js
This warning is caused by two manual mock files with the same file name.
Jest will use the mock file found in:
/Users/emmenko/dev/src/lerna-changelog/src/__mocks__/GithubAPI.js
 Please delete one of the following two files:
 /Users/emmenko/dev/src/lerna-changelog/lib/__mocks__/GithubAPI.js
/Users/emmenko/dev/src/lerna-changelog/src/__mocks__/GithubAPI.js


jest-haste-map: duplicate manual mock found:
  Module name: Changelog
  Duplicate Mock path: /Users/emmenko/dev/src/lerna-changelog/src/__mocks__/Changelog.js
This warning is caused by two manual mock files with the same file name.
Jest will use the mock file found in:
/Users/emmenko/dev/src/lerna-changelog/src/__mocks__/Changelog.js
 Please delete one of the following two files:
 /Users/emmenko/dev/src/lerna-changelog/lib/__mocks__/Changelog.js
/Users/emmenko/dev/src/lerna-changelog/src/__mocks__/Changelog.js
```

hzoo replied:
now I wonder how our jest + babel projects are doing it then? @cpojer if have ideas

cpojer replied:
We don't publish Jest's mocks to npm. They are internal to Jest, I suggest you do the same.

emmenko replied:
I also found [this thread](https://github.com/facebook/jest/issues/2070#issuecomment-265027510) from you @cpojer where you mention

```
Manual mocking is really messed up and doesn't work well.
```

Do you still recommend to use them or to "wait until they're fixed"?

Besides the publishing part, what about compiled sources from babel? Should we manually ignore those or is there a better way of doing it?
Thanks for the help!

cpojer replied:
I recommend using them within one project but I do not recommend publishing them to npm.

emmenko commented on src/Changelog.js:
```diff
@@ -1,13 +1,20 @@
+import execSync from "./execSync";
 import ConfigurationError from "./ConfigurationError";
 
+const UNRELEASED_TAG = "___unreleased___";
```
This is used for grouping unreleased commits

emmenko commented on src/Changelog.js:
```diff
@@ -1,13 +1,20 @@
+
+    // CLI options
+    this.tagFrom = options.tagFrom;
+    this.tagTo = options.tagTo;
```
Those are the new CLI options

emmenko commented on src/Changelog.js:
```diff
@@ -28,107 +35,141 @@ export default class Changelog {
-
+    // Get all info about commits in a certain tags range
+    const commitsInfo = this.getCommitsInfo();
+    const commitsByTag = this.getCommitsByTag(commitsInfo);
```
This is the new function to group commits by tag

emmenko commented on src/Changelog.js:
```diff
@@ -28,107 +35,141 @@ export default class Changelog {
+                "))";
+            });
+          });
+        });
```
Nothing changed much here, just formatting

emmenko commented on src/Changelog.js:
```diff
@@ -28,107 +35,141 @@ export default class Changelog {
+        }
+        return acc;
+      }, {})
+    );
```
Same as before, just extracted into a separate function

emmenko commented on src/Changelog.js:
```diff
@@ -28,107 +35,141 @@ export default class Changelog {
+    const commits = execSync(
+      // Prints "<short-hash>;<ref-name>;<summary>;<date>"
+      // This format is used in `getCommitsInfo` for easily analize the commit.
+      'git log --oneline --pretty="%h;%D;%s;%cd" --date=short ' + tagsRange
```
This is the most important part. Besides passing a custom range, the output is formatted with a specific pattern. This is necessary in order to analyse the commit.

```
<short-hash>;<ref-name>;<summary>;<date>
```

emmenko commented on src/Changelog.js:
```diff
@@ -28,107 +35,141 @@ export default class Changelog {
+          c => c === login || login.indexOf(c) > -1
+        )
+      );
+      if (login && shouldKeepCommiter && !committers[login]) {
```
This part includes the other option `ignoreCommitters`.

emmenko commented on src/Changelog.js:
```diff
@@ -142,74 +183,122 @@ export default class Changelog {
+        tag = _refs[1].replace(/tag:\s(.*?)$/, "$1").trim();
+      }
+      const message = parts[2];
+      const date = parts[3];
```
This part has been refactored to parse the different commit info, especially the tag name

hzoo commented on .eslintrc:
```diff
@@ -35,6 +35,8 @@
   },
   "env": {
+    "es6": true,
+    "jest": true,
```
for the eslint config we should just use https://github.com/babel/eslint-config-babel. I would do this in a new pr after this is merged so its easier to review

hzoo commented:
awesome stuff, there's a lot here so may take a while to review

hzoo commented:
cc @fson if curious/helpful?

emmenko commented:
> awesome stuff, there's a lot here so may take a while to review

Yeah take your time, I also left some TODOs in the PR description for extra stuff that can be addressed as well. I'm open to suggestions and improvements anyway on the take of this PR, so feel free to give all the feedback you have before we can proceed further ;)
And I'm happy to discuss / clarify anything you would like too.
Thanks

hzoo commented on README.md:
```diff
@@ -71,6 +71,20 @@ You'll need a GitHub API [personal access token](https://github.com/settings/tok
 - `repo`: Your "org/repo" on GitHub
 - `cacheDir` [optional]: A place to stash GitHub API responses to avoid throttling
 - `labels`: GitHub issue/PR labels mapped to changelog section headers
+- `ignoreCommitters` [optional]: list of commiters to ignore (exact or partial match)
```
Ah yeah I guess this could be useful for bots like greenkeeper?

emmenko replied:
Yep, exactly ;) We could also mention it in the CLI help

hzoo commented on README.md:
```diff
@@ -71,6 +71,20 @@ You'll need a GitHub API [personal access token](https://github.com/settings/tok
+
+
+  Options:
+    --tagFrom <tag>  define a custom tag to determine the lower bound of the range of commits (default: last available git tag)
```
Not a big deal but is `--tag-from` or `--tagFrom` more common for cli?

evocateur replied:
In my experience, `--kebab-case` (dashed) is more common for long flags. Most arg parsers will map them to the camelCase version automatically.

emmenko replied:
Good point, I'll change that 👍

hzoo commented on cli.js:
```diff
@@ -1,13 +1,26 @@
 
 var chalk = require("chalk");
-var lib = require(".");
+var argv = require("yargs").argv;
```
We don't have to but looks like we use meow in lerna already https://github.com/lerna/lerna/blob/master/bin/lerna.js#L6

evocateur replied:
I actually wanted to refactor lerna's (well, at the time, asini's) usage of meow into yargs, it's got so many neat bells and whistles (with better help out of the box).

hzoo commented on package.json:
```diff
@@ -7,9 +7,10 @@
+    "build": "npm run clean && babel src --out-dir lib --ignore src/__mocks__/GithubAPI.js,src/__mocks__/Changelog.js",
     "clean": "rimraf lib",
-    "test": "eslint index.js cli.js src/",
+    "lint": "eslint index.js cli.js src/",
```
I don't think you need the `/` for src

hzoo commented on src/Changelog.js:
```diff
@@ -142,74 +183,122 @@ export default class Changelog {
         response.commitSHA = sha;
         response.mergeMessage = message;
-        return response;
+        Object.assign(commitInfo, response);
```
This reminds me that if we have have tests we should setup travis with node 0.10 since we haven't dropped it yet. Then we can drop < node 4 in a major.

emmenko replied:
The travis build is one of the TODOs yes. Which versions do you want to test? I guess:
- `0.10`
- `0.12`
- `4`
- `6`

hzoo replied:
Yeah that sounds good.

I figured we can do a release with this and then after do a major bump to remove it

emmenko replied:
Btw, do you want to automate the release process (e.g. semantic-release) or still keep it manual (`npm publish`)? At least I would consider having a script to bump the version, tag it and publish it.
But I'm not sure how you guys are used to work and prefer to do this sorts of stuff, so yeah just asking... :)

hzoo replied:
I haven't used semantic release and most projects like Babel I would want to do a batch of commits rather than a new release per PR. But I guess for something smaller we could like this. I don't know?

emmenko replied:
No worries, I don't have much of a preference either. I would probably keep it consistent with what `lerna` is using.
Anyway, not really in the scope of this PR :)

hzoo commented on test/__snapshots__/Changelog.spec.js.snap:
```diff
@@ -0,0 +1,67 @@
+#### :rocket: New Feature
+* \`pkg-1\`, \`pkg-2\`
+  * [#1](https://github.com/lerna/lerna-changelog/pull/1) This is the commit title for the issue (#1). ([@test-user](https://github.com/test-user))
+  * [#1](https://github.com/lerna/lerna-changelog/pull/1) This is the commit title for the issue (#1). ([@test-user](https://github.com/test-user))
```
why does this have 2 of the same commit?

emmenko replied:
Yeah, actually it's correct to have 2 commit entries here. The problem is that the mock returns the same value, thus you see the same message. But I want to refactor a bit more the way we're mocking stuff, making it easier to setup the tests.
I'll push some changes soon.

hzoo replied:
Cool! Would be awesome to have a test for 2 tags (New feature/bugfix or something) and see the formatting

basically just the places where the formatting might be different

hzoo commented on test/__snapshots__/Changelog.spec.js.snap:
```diff
@@ -0,0 +1,67 @@
+
+#### Committers: 0
+"
+`;
```
If the tag is empty should it not print (same if no committers?)

emmenko replied:
Good point, I'll look into it 👍

emmenko replied:
Done https://github.com/lerna/lerna-changelog/pull/31/commits/f321146693563dc7abac051e93022b4e0d5a23a1

evocateur commented on cli.js:
```diff
@@ -1,13 +1,26 @@
 
-var Changelog = lib.Changelog;
-var ConfigurationError = lib.ConfigurationError;
+if (argv.help) {
```
If you _are_ going to use `yargs`, seems like you could do some minimal [configuration](https://github.com/yargs/yargs#optionskey-opt) and get this [`--help`](https://github.com/yargs/yargs#help) output for free (as well as `help` subcommand).

```js
var argv = require("yargs")
  .usage("$0 [options]")
  .options({
    "tag-from": {
      type: "string",
      desc: "A git tag that determines the lower bound of the range of commits (defaults to last available)"
    },
    "tag-to": {
      type: "string",
      desc: "A git tag that determines the upper bound of the range of commits"
    }
  })
  .help()
  .argv;
```

Given the powers of yargs, you might even consider the tagFrom and tagTo to be [positional arguments](https://github.com/yargs/yargs#positional-arguments) rather than flags, given that it really only makes sense to specify them in the order one might expect when specifying the range in git (`${fromTag}..${toTag}`).

emmenko replied:
Ah that's perfect! For some reasons I didn't go through the entire readme they have. I'll change that, thanks!

evocateur replied:
It is a _ridiculously_ long README, that is true. :)

emmenko replied:
Done https://github.com/lerna/lerna-changelog/pull/31/commits/ecb85202af329d93ac4ef9257f88d6d9e59f8303

The _positional arguments_ as far as I can see can only be used for a `command` which in our case is not necessary.

emmenko replied:
@evocateur so I tried the "positional arguments approach". It's kind of a workaround because `yargs` supports that only for _commands_ and because if we want them to be `<fromTag>..<toTag>` it's considered as 1 argument (because of the `..`) and we need to do some parsing anyway.

Those will be the changes. I'll keep them stashed locally until we decide whether to go with that or not.

```diff
diff --git cli.js cli.js
index 841e66a..ae2d5c0 100755
--- cli.js
+++ cli.js
@@ -1,39 +1,57 @@
 #!/usr/bin/env node
-
 var chalk = require("chalk");
 var Changelog = require(".").Changelog;
 var ConfigurationError = require(".").ConfigurationError;

 var argv = require("yargs")
-  .usage("Usage: lerna-changelog [options]")
-  .options({
-    "tag-from": {
-      type: "string",
-      desc: "A git tag that determines the lower bound of the range of commits (defaults to last available)"
-    },
-    "tag-to": {
-      type: "string",
-      desc: "A git tag that determines the upper bound of the range of commits"
-    }
-  })
+  .usage("Usage: lerna-changelog <fromTag>..<toTag>")
   .example(
     "lerna-changelog",
     "create a changelog for the changes after the latest available tag"
   )
   .example(
-    "lerna-changelog --tag-from 0.1.0 --tag-to 0.3.0",
+    "lerna-changelog 0.1.0",
+    "create a changelog for the changes after the given tag"
+  )
+  .example(
+    "lerna-changelog 0.1.0..0.3.0",
     "create a changelog for the changes in all tags within the given range"
   )
   .version()
   .help()
   .argv;

+var fromTag;
+var toTag;
+var args = argv._;
+if (args.length > 1) {
+  console.error(
+    chalk.red("Invalid arguments, must be in format `<fromTag>..<toTag>`")
+  );
+  process.exit(1);
+} else if (args.length === 1) {
+  var indexOfSeparator = args[0].indexOf("..");
+  if (indexOfSeparator === 0) {
+    console.error(
+      chalk.red("The lower bound tag is required `<fromTag>..<toTag>`")
+    );
+    process.exit(1);
+  } else if (indexOfSeparator === -1) {
+    fromTag = args[0];
+  } else {
+    var parts = args[0].split("..");
+    fromTag = parts[0];
+    toTag = parts[1];
+  }
+}
+
+var options = { fromTag: fromTag, toTag: toTag };
 try {
-  console.log((new Changelog(argv)).createMarkdown());
+  console.log(new Changelog(options).createMarkdown());
 } catch (e) {
   if (e instanceof ConfigurationError) {
     console.log(chalk.red(e.message));
   } else {
-    throw (e);
+    throw e;
   }
 }
```

hzoo commented:
This is awesome btw - although it's kind of feels like a rewrite all the code I had was pretty messy. It's difficult to really review but since there's tests that's all that matters in a sense 😄 . After all just a changelog generator

gigabo commented on src/ApiDataCache.js:
```diff
@@ -4,30 +4,32 @@ import mkdirp from "mkdirp";
-    } catch (e) {
-      // Pass.
-    }
+    } catch (e) {}
```
Was this violating some linter rule?

emmenko replied:
Not really, I just used auto-formatting from [prettier](https://github.com/jlongster/prettier).
Happy to revert it if necessary.

gigabo replied:
Oh.  That sort of thing is generally better to do in a dedicated PR.  Makes it easier to pick out relevant changes.

hzoo replied:
Yeah this is similar to the eslint comment I made above - we should just use the babel eslint config or w.e separately

emmenko replied:
I know, my bad. I usually also prefer to do it in a separate PR. I'll revert those changes, no problem

emmenko replied:
Reverted extra changes here https://github.com/lerna/lerna-changelog/pull/31/commits/a8ef94b180877f7c3c1e6c47714628521bb5fc9b

emmenko commented:
> This is awesome btw - although it's kind of feels like a rewrite all the code I had was pretty messy.
Yeah sorry, I really didn't want to rewrite/change much. But really there is not much changed, most of it it's just formatting and moving stuff around a bit. I've also added comments to mention the important points and if you feel like you need some clarification just let me know 😉

> It's difficult to really review but since there's tests that's all that matters in a sense
Exactly 😄

hzoo commented:
I think the less stuff the better or we split it out. It would be nice to merge the main change (tag from/to), other stuff can be done right after in another pr

hzoo commented:
Also invited both of you as collabs for this repo (@emmenko, @evocateur), although normally you would do it after a contribution 😛  might as well

evocateur commented:
@hzoo Thanks, I appreciate it!

> On Jan 18, 2017, at 07:14, Henry Zhu <notifications@github.com> wrote:
> 
> Also invited both of you as collabs for this repo (@emmenko, @evocateur), although normally you would do it after a contribution 😛 might as well
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

emmenko commented on src/GithubAPI.js:
```diff
@@ -36,6 +36,6 @@ export default class GithubAPI {
     }[type];
     const url = "https://api.github.com" + path;
-    return execSync("curl -H 'Authorization: token " + process.env.GITHUB_AUTH + "' --silent " + url)
+    return execSync("curl -H 'Authorization: token " + process.env.GITHUB_AUTH + "' --silent --globoff " + url)
```
This is to allow fetching url that may contain `[ ]` (e.g. `greenkeeper[bot]`)

http://stackoverflow.com/questions/8333920/passing-a-url-with-brackets-to-curl

evocateur commented:
Looking great so far!

emmenko commented:
Alright guys, I think we can do another round of review. The current status now is:
- CLI options are now `--tag-from`, `--tag-to`
---
<details>
<summary>There is also a slightly different approach by using "positional arguments"</summary>
It's kind of a workaround because `yargs` supports that only for _commands_ and because if we want them to be `<fromTag>..<toTag>` it's considered as 1 argument (because of the `..`) and we need to do some parsing anyway.

```bash
$ lerna-changelog <fromTag>..<toTag>
```

```diff
diff --git cli.js cli.js
index 841e66a..ae2d5c0 100755
--- cli.js
+++ cli.js
@@ -1,39 +1,57 @@
 #!/usr/bin/env node
-
 var chalk = require("chalk");
 var Changelog = require(".").Changelog;
 var ConfigurationError = require(".").ConfigurationError;

 var argv = require("yargs")
-  .usage("Usage: lerna-changelog [options]")
-  .options({
-    "tag-from": {
-      type: "string",
-      desc: "A git tag that determines the lower bound of the range of commits (defaults to last available)"
-    },
-    "tag-to": {
-      type: "string",
-      desc: "A git tag that determines the upper bound of the range of commits"
-    }
-  })
+  .usage("Usage: lerna-changelog <fromTag>..<toTag>")
   .example(
     "lerna-changelog",
     "create a changelog for the changes after the latest available tag"
   )
   .example(
-    "lerna-changelog --tag-from 0.1.0 --tag-to 0.3.0",
+    "lerna-changelog 0.1.0",
+    "create a changelog for the changes after the given tag"
+  )
+  .example(
+    "lerna-changelog 0.1.0..0.3.0",
     "create a changelog for the changes in all tags within the given range"
   )
   .version()
   .help()
   .argv;

+var fromTag;
+var toTag;
+var args = argv._;
+if (args.length > 1) {
+  console.error(
+    chalk.red("Invalid arguments, must be in format `<fromTag>..<toTag>`")
+  );
+  process.exit(1);
+} else if (args.length === 1) {
+  var indexOfSeparator = args[0].indexOf("..");
+  if (indexOfSeparator === 0) {
+    console.error(
+      chalk.red("The lower bound tag is required `<fromTag>..<toTag>`")
+    );
+    process.exit(1);
+  } else if (indexOfSeparator === -1) {
+    fromTag = args[0];
+  } else {
+    var parts = args[0].split("..");
+    fromTag = parts[0];
+    toTag = parts[1];
+  }
+}
+
+var options = { fromTag: fromTag, toTag: toTag };
 try {
-  console.log((new Changelog(argv)).createMarkdown());
+  console.log(new Changelog(options).createMarkdown());
 } catch (e) {
   if (e instanceof ConfigurationError) {
     console.log(chalk.red(e.message));
   } else {
-    throw (e);
+    throw e;
   }
 }
```
</details>

---
- extra formatting changes are not in this PR anymore (we can do that in a separate PR)
- I've refactored how to setup the test data in the mocks, which basically includes mainly `execSync` and `ApiDataCache`, and allows you to customize everything you need in terms of test data.
- there is now a sort of functional / integration test for the main method `createMarkdown`, I've put this in a "functional" test subfolder and uses snapshot testing. I also took the liberty of adding useful (_star wars_ related) test data 😜

/cc @hzoo @gigabo @evocateur

emmenko commented:
<details>
<summary>This is the changelog output for the full test ✍️ 🤓 </summary>

## Unreleased (2099-01-01)

#### :rocket: New Feature
* `the-force-awakens`, `rogue-one`
  * [#7](https://github.com/lerna/lerna-changelog/pull/7) feat: that is not how the Force works!. ([@han-solo](https://github.com/han-solo))

#### :nail_care: Enhancement
* `the-force-awakens`, `rogue-one`
  * [#7](https://github.com/lerna/lerna-changelog/pull/7) feat: that is not how the Force works!. ([@han-solo](https://github.com/han-solo))

#### Committers: 1
- Han Solo ([han-solo](https://github.com/han-solo))


## v6.0.0 (1983-05-25)

#### :rocket: New Feature
* `return-of-the-jedi`
  * [#5](https://github.com/lerna/lerna-changelog/pull/5) feat: I am your father. ([@vader](https://github.com/vader))

#### :bug: Bug Fix
* `return-of-the-jedi`
  * [#4](https://github.com/lerna/lerna-changelog/pull/4) fix: RRRAARRWHHGWWR. ([@chewbacca](https://github.com/chewbacca))

#### :nail_care: Enhancement
* `return-of-the-jedi`
  * [#6](https://github.com/lerna/lerna-changelog/pull/6) refactor: he is my brother. ([@princess-leia](https://github.com/princess-leia))

#### :house: Maintenance
* `return-of-the-jedi`
  * [#4](https://github.com/lerna/lerna-changelog/pull/4) fix: RRRAARRWHHGWWR. ([@chewbacca](https://github.com/chewbacca))

#### Committers: 3
- Chwebacca ([chewbacca](https://github.com/chewbacca))
- Darth Vader ([vader](https://github.com/vader))
- Princess Leia Organa ([princess-leia](https://github.com/princess-leia))


## v5.0.0 (1980-05-17)

#### :boom: Breaking Change
* `empire-strikes-back`
  * [#2](https://github.com/lerna/lerna-changelog/pull/2) chore: Terminate her... immediately!. ([@gtarkin](https://github.com/gtarkin))

#### :bug: Bug Fix
* `empire-strikes-back`
  * [#3](https://github.com/lerna/lerna-changelog/pull/3) fix: Get me the rebels base!. ([@vader](https://github.com/vader))

#### Committers: 2
- Darth Vader ([vader](https://github.com/vader))
- Governor Tarkin ([gtarkin](https://github.com/gtarkin))


## v4.0.0 (1977-05-25)

#### :rocket: New Feature
* `a-new-hope`
  * [#1](https://github.com/lerna/lerna-changelog/pull/1) feat: May the force be with you. ([@luke](https://github.com/luke))

#### Committers: 1
- Luke Skywalker ([luke](https://github.com/luke))"
</details>

hzoo commented:
Very cool 👍

emmenko commented:
@hzoo @gigabo @evocateur sorry guys, I know we are all busy with more important stuff, just wondering if you got a chance to have another look at this. Thanks ;)

hzoo commented:
I think it's good we can continue to make new prs to make it better. Difficult to review otherwise.

hzoo commented on test/functional/__snapshots__/markdown-full.spec.js.snap:
```diff
@@ -0,0 +1,64 @@
+#### Committers: 1
+- Han Solo ([han-solo](https://github.com/han-solo))
+
+
```
Non-blocking: wondering why there's 2 new lines here (unless we want that)

emmenko replied:
I thought it's a bit better to distinguish between tags / releases. If you think it's not important I can put only 1 break like back.

hzoo commented on test/functional/__snapshots__/markdown-full.spec.js.snap:
```diff
@@ -0,0 +1,64 @@
+exports[`createMarkdown outputs correct changelog 1`] = `
+"
```
Is the `"` here intentional + the escaped quotes below?

emmenko replied:
Yeah, it's part of the snapshot.

```js
describe('hello', () => {
  it('world', () => {
    expect('hello\nworld').toMatchSnapshot()
  })
})

exports[`foo bar 1`] = `
"hello
world"
`;
```

evocateur commented on package.json:
```diff
@@ -7,9 +7,10 @@
   },
   "scripts": {
-    "build": "babel src -d lib",
+    "build": "npm run clean && babel src --out-dir lib --ignore src/__mocks__/GithubAPI.js,src/__mocks__/Changelog.js,src/__mocks__/ApiDataCache.js,src/__mocks__/execSync.js",
```
Wouldn't `--ignore src/__mocks__` be sufficient here?

emmenko replied:
True, thanks 😄

hzoo commented on test/functional/__snapshots__/markdown-full.spec.js.snap:
```diff
@@ -0,0 +1,64 @@
+## Unreleased (2099-01-01)
+
+#### :rocket: New Feature
+* \`the-force-awakens\`, \`rogue-one\`
```
I don't think we want it to be outputting ``` \` ``` in the changelog if that's possible?

emmenko replied:
It's just what the snapshot output looks like, the normal output doesn't have the escapes.

hzoo replied:
Oh ok 😄

hzoo commented:
lgtm just wondering about the escaped quotes in the changelog output

hzoo commented:
From master on Babel

<details>
## Unreleased (2017-01-26)

#### :bug: Bug Fix
* `babel-generator`
  * [#5193](https://github.com/babel/babel/pull/5193) Fix missing parens when function expressions is tag. ([@existentialism](https://github.com/existentialism))

#### :memo: Documentation
* `babel-plugin-transform-flow-strip-types`
  * [#5212](https://github.com/babel/babel/pull/5212) Remove REPL link transform-flow-strip-types doc. ([@xtuc](https://github.com/xtuc))
* `babel-plugin-transform-regenerator`
  * [#5202](https://github.com/babel/babel/pull/5202) Fix transform-regenerator README. ([@xtuc](https://github.com/xtuc))
* `babel-plugin-transform-es2015-arrow-functions`
  * [#5200](https://github.com/babel/babel/pull/5200) Fix transform-es2015-arrow-functions code blocks on the website. ([@xtuc](https://github.com/xtuc))
  * [#5194](https://github.com/babel/babel/pull/5194) Fix transform-es2015-arrow-functions README. ([@xtuc](https://github.com/xtuc))

#### :house: Internal
* Other
  * [#5182](https://github.com/babel/babel/pull/5182) Run coverage only once. ([@existentialism](https://github.com/existentialism))
  * [#5165](https://github.com/babel/babel/pull/5165) Add Node 7 to CI. ([@chicoxyzzy](https://github.com/chicoxyzzy))

#### Committers: 4
- Brian Ng ([existentialism](https://github.com/existentialism))
- Henry Zhu ([hzoo](https://github.com/hzoo))
- Sergey Rubanov ([chicoxyzzy](https://github.com/chicoxyzzy))
- Sven SAULEAU ([xtuc](https://github.com/xtuc))
</details>

emmenko commented:
One last thing that I forgot guys (see https://github.com/lerna/lerna-changelog/pull/31/commits/0408bcba1cb7a9e89e582de87e76418707ac522c): in case there are **multiple** tags referenced by the same commit, we should group each one of them separately (each one will have the same list of commits).
In my opinion this is the _expected behaviour_, it makes it also easier to copy-paste the log for each tag in their respective github release page.

Example:

```bash
$ git log --oneline --pretty="%h;%D;%s;%cd" --date=short v0.1.0..
a0000004;tag: a-new-hope@4.0.0, tag: empire-strikes-back@5.0.0, tag: return-of-the-jedi@6.0.0;chore(release): releasing component;1977-05-25
a0000003;;Merge pull request #1 from star-wars;1977-05-25
a0000002;tag: v0.1.0;chore(release): releasing component;1966-01-01
a0000001;;fix: some random fix which will be ignored;1966-01-01
```

![image](https://cloud.githubusercontent.com/assets/1110551/22403350/88b9252e-e614-11e6-9087-527a3cfbdec3.png)

---

Alternatively such multiple tags are listed in the same group of commits, but I find it kind of wrong.

![image](https://cloud.githubusercontent.com/assets/1110551/22403359/b9cd767e-e614-11e6-8575-220c5025226a.png)

emmenko commented:
On the other hand, if I merge different kind of PRs referring to different packages (with different changes), then do a release round, lerna will publish all different packages and their related tags. If we generate the changelog at this point in time, we'll result having this "huge" group of tags containing all sorts of commits, mostly unrelated to each other, which is kind of confusing.

Oh man, I'm getting really confused myself now with how exactly the publishing / tagging should work for the different packages 😵

Any idea how we should approach this?
I guess this is a problem more related to packages using independent versioning and tags named after the packages themselves.

hzoo commented:
> I guess this is a problem more related to packages using independent versioning and tags named after the packages themselves.

Yes! I think for both modes we'd want to bundle up the changelog per release rather than by tag. So need to differentiate between multiple tag releases at once in independent mode vs just trying to print out previous versions (which shouldn't happen that often really)

emmenko commented:
Can you give some examples how you would expect this to be?

hzoo commented:
Thinking about cra (https://github.com/facebookincubator/create-react-app/releases) - the way lerna works currently is that it makes a tag for each pkg. I mean maybe you want a separate heading for each release/pkg but I don't know. I'm ok with this for now? We can just make it a major version and do some patches

hzoo commented:
We can always push out some patches

Feel free to make issues for the things we didn't get to

hzoo commented:
@emmenko trying to figure out how to run `lerna-changelog` on itself given this PR 😛

hzoo commented:
Thanks so much for this awesome work btw! (Wanted to get a release out but it made sense to try out a changelog on this repo since ironically it doesn't have one itself)

emmenko commented:
Glad I could help somehow! Hope to figure out soon this release/tag thingy. 🤔

hzoo commented:
@emmenko can you make issues for the things you missed before?

emmenko commented:
@hzoo most of the stuff has been already done (eslint, travis, etc). Should be fine then, if I think of something else I'll open an issue
